ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs

```diff
@@ -0,0 +1,364 @@
+            fullType <- fullType.GenericArguments[1]
+
+        result
+
+    // methods take "this" arg as first parameter so skip it
+    let getMethodParameterTypes parameters (fcsType: FSharpType) =
+        let result = Array.zeroCreate parameters
+        let mutable fullType = fcsType.GenericArguments[1]
+
+        for i = 0 to parameters - 1 do
+            result[i] <- fullType.GenericArguments[0]
+            fullType <- fullType.GenericArguments[1]
+
+        result
+
+    let getFunctionParameterAt parameterIndex (fcsType: FSharpType) =
+        let rec getParameter index (fcsType: FSharpType) =
+            if index = 0 then fcsType.GenericArguments[0]
+            else
+                getParameter (index - 1) fcsType.GenericArguments[1]
+
+        getParameter parameterIndex fcsType
+
+module AnnotationUtil =
+
+    let private isFullyAnnotatedNamedTypeUsage (namedTypeUsage: INamedTypeUsage) =
+        isNull namedTypeUsage.ReferenceName ||
+        isNull namedTypeUsage.ReferenceName.TypeArgumentList ||
+        namedTypeUsage.ReferenceName.TypeArgumentList.TypeUsagesEnumerable
+           |> Seq.exists (fun typeUsage -> typeUsage :? IAnonTypeUsage)
+           |> not
+
+    let rec private isFullyAnnotatedFunctionTypeUsage (functionTypeUsage: IFunctionTypeUsage) =
+        isFullyAnnotatedTypeUsage functionTypeUsage.ArgumentTypeUsage &&
+        isFullyAnnotatedTypeUsage functionTypeUsage.ReturnTypeUsage
+
+    and isFullyAnnotatedTypeUsage (typeUsage: ITypeUsage) =
+        match typeUsage with
+        | :? INamedTypeUsage as namedTypeUsage ->
+            isFullyAnnotatedNamedTypeUsage namedTypeUsage
+        | :? IFunctionTypeUsage as functionTypeUsage ->
+            isFullyAnnotatedFunctionTypeUsage functionTypeUsage
+        | :? IConstrainedTypeUsage as constrainedTypeUsage ->
+            isFullyAnnotatedTypeUsage constrainedTypeUsage.TypeUsage
+        | _ -> false
+
+    let isFullyAnnotatedReturnTypeInfo (returnTypeInfo: IReturnTypeInfo) =
+        isNotNull returnTypeInfo.ReturnType &&
+        isFullyAnnotatedTypeUsage returnTypeInfo.ReturnType
+
+    let rec isFullyAnnotatedPattern (pattern: IFSharpPattern) =
+        match pattern.IgnoreInnerParens() with
+        | :? IUnitPat ->
+            true
+        | :? ITypedPat as typedPat ->
+            isFullyAnnotatedPattern typedPat.Pattern
+        | pattern ->
+            pattern |> PatUtil.getTypeUsage |> ValueOption.exists isFullyAnnotatedTypeUsage
+            || pattern |> PatUtil.getReturnTypeInfo |> ValueOption.exists isFullyAnnotatedReturnTypeInfo
+
+    let isFullyAnnotatedBinding (binding: IBinding) =
+        isNotNull binding.ReturnTypeInfo &&
+        binding.ParametersDeclarationsEnumerable
+        |> Seq.forall (fun parameter ->
+            isFullyAnnotatedPattern parameter.Pattern)
+
+    let isFullyAnnotatedMemberDeclaration (memberDeclaration: IMemberDeclaration) =
+        isNotNull memberDeclaration.ReturnTypeInfo &&
```

Comment:
Maybe this check could be moved into `isFullyAnnotatedReturnTypeInfo` to be consistent with `isFullyAnnotatedNamedTypeUsage`.

Reply:
Moved